### PR TITLE
[C#] Support tuples whose commas are not on the same line as the open paren

### LINKED ISSUE #2832: [C#] Method expression-bodies members
- Sublime Version: 3.2.2, Build 3211
- OS Version: Windows 10 19042.985

Hi, I am seeing this issue in C# where method expressions show a syntax error on deeply-nested code.

``` csharp
[Fact]
public void ApplyValidationFail() =>
  Assert.Equal(
    "[[fail_1], [fail_2]]",
    (
      Success<Seq<string>, Option<int>>(3),
      Fail<Seq<string>, Option<int>>(Seq1("fail_1")),
      Success<Seq<string>, Option<int>>(3), Fail<Seq<string>,
      Option<int>>(Seq1("fail_2"))
    )
    .Apply((a, b, c, d) =>
      from w in a
      from x in b
      from y in c
      from z in d
      select w + x + y + z)
    .FailToSeq()
    .ToString()
  ); // <-- Syntax error
```

Seems to work fine on simple expressions.

```chsarp
[Fact]
public void BindProp() =>
  Assert.Equal(
    None,
    Optional(1)
      .Bind<int>(o => o + 2)
      .Bind<int>(_ => None)
      .Bind<int>(_ => 1)
  ); // <-- No syntax error
```

Attached are screenshots of how this shows in Sublime.

image
image

### PR DESCRIPTION
**Pull request by keith-hall**
Fixes #2832

### CODE REVIEW

File: C#/C#.sublime-syntax [hunk 1]
```diff
@@ -1377,6 +1346,49 @@ contexts:
     - match: (?=\}|\)|>|\]|,|{{reserved}})
       pop: true
 
+  tuple:
+    - meta_scope: meta.group.tuple.cs
```
Comment (deathaxe) [inline C#/C#.sublime-syntax:1350]: Would we probably want to use `meta.sequence.tuple` and corresponding punctuations to line up with python?

File: C#/C#.sublime-syntax [hunk 2]
```diff
@@ -1330,43 +1330,12 @@ contexts:
...
     - match: \(
       scope: punctuation.section.group.begin.cs
-      push:
-        - meta_scope: meta.group.cs
-        - match: \)
-          scope: punctuation.section.group.end.cs
-          pop: true
-        - include: lambdas
-        - include: line_of_code_in_no_semicolon
+      branch_point: tuple_or_group
```
Comment (deathaxe) [inline C#/C#.sublime-syntax:1335]: I'd probably vote for using the same name scheme as for contexts and keep underscore use reserved for variables only. It makes it easier to give them the same name as a context, as already noted in another PR.

Comment (deathaxe) [reply C#/C#.sublime-syntax:1335]: Nevermind. I just see this syntax uses underscors for contexts, too. So this name is very fine.

File: C#/C#.sublime-syntax [hunk 3]
```diff
@@ -1377,6 +1346,49 @@ contexts:
...
+      scope: keyword.operator.assignment.cs
+    - match: _\b
+      scope: variable.language.deconstruction.discard.cs
+    - match: (?!{{reserved}})(?={{namespaced_name}}{{type_suffix}}\s+{{name}}\s*[:,])
+      push: var_declaration_explicit
+    - match: '({{name}})(<)'
+      captures:
+        1: support.type.cs
+        2: meta.generic.cs punctuation.definition.generic.begin.cs
+      push:
```
Comment (deathaxe) [inline C#/C#.sublime-syntax:1366]: We probably should try to introduce "named contexts for everything". Not sure if C# is a candidate, but it helps with inheriting syntaxes as well as debugging via scope popup, which now shows context names.

Comment (keith-hall) [reply C#/C#.sublime-syntax:1366]: I like this point. May be worth adding to https://github.com/sublimehq/Packages/issues/757

Comment (deathaxe) [reply C#/C#.sublime-syntax:1366]: I wonder if we should create a more structured kind of documentation for things like that in the long term.

I have too much such details in mind but somehow find it hard to create documentation from it.

File: C#/C#.sublime-syntax (hunk 1 above)
Comment (keith-hall) [reply C#/C#.sublime-syntax:1350]: I agree. I took a look into this, and it's a complicated change. Maybe I can do it as part of a separate PR please? :)

Comment (deathaxe) [reply C#/C#.sublime-syntax:1350]: Sure.

File: C#/C#.sublime-syntax (hunk 3 above)
Comment (deathaxe) [reply C#/C#.sublime-syntax:1366]: Refactoring syntaxes to use named contexts can be done in dedicated PRs as well.

**Review by deathaxe — APPROVED** [review]

**Review by FichteFoll — APPROVED** [review]

**Pull request merged**